mat64/list_test.go

```diff
@@ -0,0 +1,684 @@
+	denseComparison func(receiver, a, b *Dense),
+
+	// legalTypes returns whether the concrete types in Matrix are valid for
+	// the method.
+	legalTypes func(a, b Matrix) bool,
+
+	// dimsOK returns whether the matrix sizes are valid for the method.
+	dimsOK func(ar, ac, br, bc int) bool,
+) {
+	strideVec := &Vector{
+		mat: blas64.Vector{
+			Inc: 10,
+		},
+	}
+	// Loop over all of the matrix types.
+	types := []Matrix{
+		&Dense{},
+		&SymDense{},
+		NewTriDense(0, true, nil),
+		NewTriDense(0, false, nil),
+		NewVector(0, nil),
+		strideVec,
+		&basicMatrix{},
+		&basicVectorer{},
+		&basicSymmetric{},
+		&basicTriangular{},
+	}
+
+	for _, aMat := range types {
+		for _, bMat := range types {
+			// Loop over all of the size combinations (bigger, smaller, etc.).
+			for _, test := range []struct {
+				ar, ac, br, bc int
+			}{
+
+				{1, 1, 1, 1},
+				{6, 6, 6, 6},
+				{7, 7, 7, 7},
+
+				{1, 1, 1, 5},
+				{1, 1, 5, 1},
+				{1, 5, 1, 1},
+				{5, 1, 1, 1},
+				{6, 6, 6, 11},
+				{6, 6, 11, 6},
+				{6, 11, 6, 6},
+				{11, 6, 6, 6},
+				{11, 11, 11, 6},
+				{11, 11, 6, 11},
+				{11, 6, 11, 11},
+				{6, 11, 11, 11},
+
+				{1, 1, 5, 5},
+				{1, 5, 1, 5},
+				{1, 5, 5, 1},
+				{5, 1, 1, 5},
+				{5, 1, 5, 1},
+				{5, 5, 1, 1},
+				{6, 6, 11, 11},
+				{6, 11, 6, 11},
+				{6, 11, 11, 6},
+				{11, 6, 6, 11},
+				{11, 6, 11, 6},
+				{11, 11, 6, 6},
+
+				{1, 1, 17, 11},
+				{1, 1, 11, 17},
+				{1, 11, 1, 17},
+				{1, 17, 1, 11},
+				{1, 11, 17, 1},
+				{1, 17, 11, 1},
+				{11, 1, 1, 17},
+				{17, 1, 1, 11},
+				{11, 1, 17, 1},
+				{17, 1, 11, 1},
+				{11, 17, 1, 1},
+				{17, 11, 1, 1},
+
+				{6, 6, 1, 11},
+				{6, 6, 11, 1},
+				{6, 11, 6, 1},
+				{6, 1, 6, 11},
+				{6, 11, 1, 6},
+				{6, 1, 11, 6},
+				{11, 6, 6, 1},
+				{1, 6, 6, 11},
+				{11, 6, 1, 6},
+				{1, 6, 11, 6},
+				{11, 1, 6, 6},
+				{1, 11, 6, 6},
+
+				{6, 6, 17, 1},
+				{6, 6, 1, 17},
+				{6, 1, 6, 17},
+				{6, 17, 6, 1},
+				{6, 1, 17, 6},
+				{6, 17, 1, 6},
+				{1, 6, 6, 17},
+				{17, 6, 6, 1},
+				{1, 6, 17, 6},
+				{17, 6, 1, 6},
+				{1, 17, 6, 6},
+				{17, 1, 6, 6},
+
+				{6, 6, 17, 11},
+				{6, 6, 11, 17},
+				{6, 11, 6, 17},
+				{6, 17, 6, 11},
+				{6, 11, 17, 6},
+				{6, 17, 11, 6},
+				{11, 6, 6, 17},
+				{17, 6, 6, 11},
+				{11, 6, 17, 6},
+				{17, 6, 11, 6},
+				{11, 17, 6, 6},
+				{17, 11, 6, 6},
+			} {
+				// Skip the test if the arguments are compile-time provably incorrect
+				// This is either because the matrix types are wrong at compile time
+				// or because the sizes are impossible.
+				if !legalTypes(aMat, bMat) {
+					continue
+				}
+				if !legalDims(aMat, test.ar, test.ac) {
+					continue
+				}
+				if !legalDims(bMat, test.br, test.bc) {
+					continue
+				}
+				a := makeRandOf(aMat, test.ar, test.ac)
+				b := makeRandOf(bMat, test.br, test.bc)
+
+				// Compute the true answer if the sizes are legal.
+				legal := dimsOK(test.ar, test.ac, test.br, test.bc)
```

Comment:
I think there was confusion here. The func should be legalDims and the status var should dimsOK. legal by itself doesn't have enough context to span this many lines.